feat: add EzLineChart component (#959)
ezcater/recipe: packages/recipe/src/components/EzLineChart/EzLineChart.md

---
name: EzLineChart
title: Line Chart
category: Data
path: '/components/ez-line-chart'
tags: ['chart', 'graph', 'bar', 'line', 'progress']
---

Line charts are a [type of chart](https://help.flourish.studio/article/25-line-bar-and-pie-charts) consisting of a line that describe a dataset.

---

## Best Practices

- Line charts are used to a changing time series
- Line charts are good for showing trends over time

---

## Examples

### Basic line chart

Line charts should include a title and container in an `EzCard` container, the title can optionally have a description and icon.

> To chart a data set as a line chart, provide a dataset as a `data` array of objects with `x` and `y` values, for example: `data = [{x: 'first', y: 1}, {x: 'second', y: 2}, {x: 'third', y: 3}]`

Reviewer: Perhaps this can be formatted to be easier to read?

---

To create a line chart, provide a data set as a `data` array of objects with `x` and `y` values. For example:

```js
data = [
  {x: 'first', y: 1},
  {x: 'second', y: 2},
  {x: 'third', y: 3},
];
```

> Always provide `title` and a detailed `description` for assistive devices.

Reviewer: I would move this to an "Accessibility" section near the bottom to match our other component docs.

Author: it's part of the required props though?

Typically, the independent axis will be the x-axis and the dependent axis will be the y-axis. To format the point labels on each axis, optionally pass a function to each `dependentAxisLabelFormatter` and/or `independentAxisLabelFormatter`. To customize which labels appear on each axis, optionally pass an array of values of the same type as the respective `data` values to each `dependentAxisLabelValues` and/or `independentAxisLabelValues`.

> Lines colors can be customized with [theme colors](http://localhost:8000/guides/theming/#colors), the default is `primary`.

Reviewer: Should this be in a separate "Line Chart Colors" section to match our other component doc pages?

```suggestion
Line colors can be customized with [theme colors](/guides/theming/#colors). The default color is `primary`.
```

> If your dataset may sometimes be an empty array, work with design to render something other than a blank chart, perhaps an `EzBlankState`.

Reviewer: ```suggestion
If your data set may sometimes be an empty array, work with design to render something other than a blank chart, perhaps an [`EzBlankState`](/components/ez-blank-state).
```

```jsx
() => {
  const DATA = [
    {x: 'January', y: 1000},
    {x: 'February', y: 1100},
    {x: 'March', y: 1500},
    {x: 'April', y: 1400},
    {x: 'May', y: 1500},
    {x: 'June', y: 1800},
    {x: 'July', y: 2000},
    {x: 'August', y: 2200},
    {x: 'September', y: 2500},
    {x: 'October', y: 2800},
    {x: 'November', y: 3000},

>     {x: 'Decemeber', y: 3500},

Reviewer: ```suggestion
    {x: 'December', y: 3500},
```

  const formatter = new Intl.NumberFormat('en-US', {
    minimumFractionDigits: 0,
    maximumFractionDigits: 0,
  });

  const independentAxisLabelFormatter = t => `${t.substring(0, 3)}`;
  const dependentAxisLabelFormatter = t => `${formatter.format(t)}`;

  const DEPENDENT_AXIS_LABEL_VALUES = [0, 1000, 2000, 3000, 4000, 5000];

  return (
    <EzPage>
      <EzCard
        title={METRIC}
        subtitle={TIME_PERIOD}
        actions={<EzHeading size="1">{SUM}</EzHeading>}
      >
        {!DATA || DATA.length === 0 ? (
          <EzBlankState
            title="No data available"
            description="We may not have enough information yet."
          />
        ) : (
          <EzLineChart
            data={DATA}
            description={description}
            dependentAxisLabelFormatter={dependentAxisLabelFormatter}
            dependentAxisLabelValues={DEPENDENT_AXIS_LABEL_VALUES}
            independentAxisLabelFormatter={independentAxisLabelFormatter}
            title={title}
          />
        )}
      </EzCard>
    </EzPage>
  );
};
```

---

## Props

```jsx-hide-controls
  <PropsTable propsData={[
    {
      name: 'data*',

>       types: ['array of datum'],

Reviewer: ```suggestion
      types: ['Datum[]'],
```

      description: 'Array of data with x and y values of type string, number, array of strings, or function.',
    },
    {
      name: 'description*',
      types: ['string'],
      description: 'Detailed accessible description for the chart.',
    },
    {
      name: 'dependentAxisLabelFormatter',

>       types: ['function'],

Reviewer: For consistency with our other component docs.
```suggestion
      types: ['func'],
```

      description: 'The function to format the dependent axis points.',
    },
    {
      name: 'dependentAxisLabelValues',

>       types: ['array of axis label values'],

Reviewer: ```suggestion
      types: ['AxisLabelValue[]'],
```

      description: 'Array of axis label values to customize the dependent axis labels.',
    },
    {
      name: 'independentAxisLabelFormatter',

>       types: ['function'],

Reviewer: For consistency with our other component docs.
```suggestion
      types: ['func'],
```

      description: 'The function to format the independent axis points.',
    },
    {
      name: 'independentAxisLabelValues',

>       types: ['array of axis label values'],

Reviewer: ```suggestion
      types: ['AxisLabelValue[]'],
```

      description: 'Array of axis label values to customize the independent axis labels.',
    },
    {
      name: 'lineColor',
      types: ['EzThemeColors'],
      defaultValue: 'primary',
      description: 'Line chart line color.',
    },
    {
      name: 'title*',
      types: ['string'],
      description: 'Accessible title for the chart.',

>     },

Reviewer: I would put any required props at the top, in alphabetical order.

Author: yeah, i wrote these in alphabetical order

Reviewer: It's common practice to split up the props with required props on the top and optional props on the bottom, both in respective alphabetical order. We do this in our other components. This is also how Storybook Docs orders them. Should we keep them consistent?

  ]}>
  </PropsTable>
```

---

## Related components

> - [Bar Chart](/components/ez-bar-chart)
> - [Progress](/components/ez-progress)

Reviewer: ```suggestion
- [EzBarChart](/components/ez-bar-chart)
- [EzProgress](/components/ez-progress)
- [EzBlankState](/components/ez-blank-state)
```

---
name: EzLineChart
title: Line Chart
category: Data
path: '/components/ez-line-chart'
tags: ['chart', 'graph', 'bar', 'line', 'progress']
---

Line charts are a type of chart consisting of a line that describe a dataset.

---

## Best Practices

> - Line charts are used to a changing time series

Author: i grabbed it from an external resource actually. i can add a link. we have best practices at the top of all our components, usually source this from designers and a mix of external resources

        fill="#00B373"
      />
    </svg>

>   );

Reviewer 2: Do we need a header icon to get our point across with this example? If we can simplify the code samples down to their bare minimums, I'd imagine we make users more comfortable using our components.
